Source/Dafny/AST/DafnyAst.cs

```diff
@@ -13,14 +13,92 @@
 using System.Diagnostics.Contracts;
 using System.Numerics;
 using System.Linq;
-using Microsoft.Boogie;
 using System.Diagnostics;
 using System.Threading;
+using Microsoft.Boogie;
 
 namespace Microsoft.Dafny {
   [System.AttributeUsage(System.AttributeTargets.Field)]
   public class FilledInDuringResolutionAttribute : System.Attribute { }
 
+  public interface IToken : Microsoft.Boogie.IToken {
+    // Inherited for now.
+    /*
+    int kind { get; set; }
+    string filename { get; set; }
+    int pos { get; set; }
+    int col { get; set; }
+    int line { get; set; }
+    string val { get; set; }
+    bool IsValid { get; }*/
+    string leadingTrivia { get; set; }
+    string trailingTrivia { get; set; }
+  }
+
+  public record Token : IToken {
+    private int _kind;
+    private string _filename;
+    private int _pos;
+    private int _col;
+    private int _line;
+    private string _val;
+    private string _leadingTrivia;
+    private string _trailingTrivia;
+    public Token next;
+    public static readonly IToken NoToken = (IToken)new Token();
```

Comment:
What's this? Shouldn't the optionality be one level higher, in the AST Nodes, to avoid having bogus values for source location related fields.

Reply:
I imagine we want to end with something like:

```c#
record TokenRange(IToken: start, IToken end)
record ASTNode(range: TokenRange?)
```

So there's no `NoToken`, generated AST nodes just use `null` instead of an instance of `TokenRange`.

Reply:
There are 220 occurrences of `Token.NoToken` in the DafnyPipeline, 39 in the parser, 21 in the resolver, 50 in the verifier, 67 in the compilers, and 36 in the AST. 
Although you might be right, I implore you to consider that removing NoToken from the source code is an implementation detail and out of scope for this PR.
Besides, there are places that actually check if a token is null or not (which is different from being a NoToken), to enable or disable some characteristics (like in the parser, if there is a token for "method" after "function", then it's a function method). I know the implementation should be better, but this PR is already touching a lot of code just to remove the dependency on Boogie's tokens until verification,.